edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -49,43 +50,26 @@ lazy_static! {
     };
 }
 
+macro_rules! get_client {
+    ($self:ident) => {
+        match $self {
+            DockerModuleRuntime::Uninitialized => {
+                return Box::new(Err(Error::from(ErrorKind::Uninitialized)).into_future());
+            }
+            DockerModuleRuntime::Initialized(ref client) => client,
+        }
+    };
+}
+
 #[derive(Clone)]
-pub struct DockerModuleRuntime {
-    client: DockerClient<UrlConnector>,
-    network_id: Option<String>,
+pub enum DockerModuleRuntime {
+    Uninitialized,
```

Comment:
If you make `ModuleRuntime::init()` the constructor that returns `Future<Self>` instead of this "async init" business, then you wouldn't need this `Uninitialized` state and the `get_client` macro.

Reply:
Makes sense.

Reply:
Addressing this required adding a `MakeModuleRuntime` trait in core representing creating a type that implements `ModuleRuntime` and removing `ModuleRuntime::init` entirely.